scalalib/worker/src/mill/scalalib/worker/ZincWorkerImpl.scala

```diff
@@ -708,4 +684,38 @@ object ZincWorkerImpl {
       toAnalyze = List((List(start), deps(start).toList))
     ).reverse
   }
+
+  /**
+   * In memory cache backed by a ConcurrentHashMap[A, SoftReference[B]]
+   * `SoftReference`s allow the GC to free values when memory is scarce
+   */
+  private class Cache[A, B <: AnyRef] {
+    protected val cache =
+      new java.util.concurrent.ConcurrentHashMap[A, SoftReference[B]]
+
+    def getOrElseCreate(a: A)(create: => B): B =
+      cache.compute(
+        a,
+        {
+          case (_, v @ SoftReference(_)) => v
+          case _ => SoftReference(create)
+        }
+      )()
```

Comment:
Isn't there a tiny risk that the instance returned by `create` (or the existing one) is garbage-collected by the time the `SoftReference` is dereferenced? Having a var to maintain a strong reference throughout the lifecycle of `getOrElseCreate` feels safer.

Reply:
Yeah, this doesn't look right.  The scheme that I used in `Soft` (which is a `SoftReference` that can recreate itself if it's needed after it's cleared) is better: https://github.com/Ichoran/kse3/blob/5a900ad9578008f0f3454c759cb83e73ec0e21a8/flow/src/Cached.scala#L105-L154